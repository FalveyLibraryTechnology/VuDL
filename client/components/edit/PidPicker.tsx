import React, { useState } from "react";
import ChildList from "./children/ChildList";
import Accordion from "@mui/material/Accordion";
import AccordionSummary from "@mui/material/AccordionSummary";
import AccordionDetails from "@mui/material/AccordionDetails";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import { useEditorContext } from "../../context/EditorContext";

export interface Parent {
    pid: string;
    title: string;
}

interface PidPickerProps {
    selected: string;
    setSelected: (pid: string) => void;
    parents?: Array<Parent>;
}

const PidPicker = ({ selected, setSelected, parents = [] }: PidPickerProps): React.ReactElement => {
    const {
        state: { favoritePidsCatalog },
    } = useEditorContext();
    const [textboxPid, setTextboxPid] = useState<string>("");
    const favorites = [];
    for (const pid in favoritePidsCatalog) {
        favorites[favorites.length] = (
            <li key={`favorite_${pid}`}>
                <button onClick={() => setSelected(pid)}>{favoritePidsCatalog[pid]}</button>
            </li>
        );
    }
    const favoritesAccordion =
        favorites.length > 0 ? (
            <Accordion>
                <AccordionSummary expandIcon={<ExpandMoreIcon />}>Choose PID from Favorites</AccordionSummary>
                <AccordionDetails>
                    <ul>{favorites}</ul>
                </AccordionDetails>
            </Accordion>
        ) : null;

    const parentAccordions = parents.map((parent: Parent) => {
        return (
            <Accordion key={`clone_parent_${parent.pid}`}>
                <AccordionSummary expandIcon={<ExpandMoreIcon />}>
                    Choose PID from Parent ({parent.title})
                </AccordionSummary>
                <AccordionDetails>
                    <ChildList pid={parent.pid} selectCallback={setSelected} />
                </AccordionDetails>
            </Accordion>
        );
    });

    return selected.length > 0 ? (
        <>
            Selected pid: {selected}. <button onClick={() => setSelected("")}>Clear</button>
        </>
    ) : (
        <>
            <Accordion>
                <AccordionSummary expandIcon={<ExpandMoreIcon />}>Enter PID manually</AccordionSummary>
                <AccordionDetails>
                    <label>
                        Enter PID:{" "}
                        <input type="text" value={textboxPid} onChange={(e) => setTextboxPid(e.target.value)} />
                    </label>
                    <button onClick={() => setSelected(textboxPid)}>Set</button>
                </AccordionDetails>
            </Accordion>
            <Accordion>
                <AccordionSummary expandIcon={<ExpandMoreIcon />}>Choose PID from Tree</AccordionSummary>
                <AccordionDetails>
                    <ChildList selectCallback={setSelected} />
                </AccordionDetails>
            </Accordion>
            {parentAccordions}
            {favoritesAccordion}
        </>
    );
};

export default PidPicker;
