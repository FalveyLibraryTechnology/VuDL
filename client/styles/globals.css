:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
