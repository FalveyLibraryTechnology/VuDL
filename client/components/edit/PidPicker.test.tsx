import React from "react";
import { describe, beforeEach, expect, it, jest } from "@jest/globals";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { fireEvent, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import PidPicker from "./PidPicker";
import { Parent } from "./PidPicker";

const mockUseEditorContext = jest.fn();
jest.mock("../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));
jest.mock("./children/ChildList", () => () => "ChildList");

describe("PidPicker", () => {
    let callback: () => void;
    let editorValues;
    let favoritePidsCatalog: Record<string, string>;

    const getPicker = (selected = "", parents: Array<Parent> = []) => {
        return <PidPicker selected={selected} parents={parents} setSelected={callback} />;
    };

    const checkSnapshot = (selected = "", parents: Array<Parent> = []) => {
        const wrapper = shallow(getPicker(selected, parents));
        expect(toJson(wrapper)).toMatchSnapshot();
    };

    beforeEach(() => {
        callback = jest.fn();
        favoritePidsCatalog = { "foo:123": "first test", "foo:124": "second test" };
        editorValues = {
            state: { favoritePidsCatalog },
        };
        mockUseEditorContext.mockReturnValue(editorValues);
    });

    it("renders correctly with favorite PIDs", () => {
        checkSnapshot();
    });

    it("sets selected PID when you click a favorite button", async () => {
        render(getPicker());
        await userEvent.setup().click(screen.getByText("first test"));

        expect(callback).toHaveBeenCalledWith("foo:123");
    });

    it("sets selected PID when you use manual entry", async () => {
        render(getPicker());
        fireEvent.change(screen.getByRole("textbox", { hidden: true }), { target: { value: "bar" } });
        await userEvent.setup().click(screen.getByText("Set"));

        expect(callback).toHaveBeenCalledWith("bar");
    });

    it("renders correctly with a selected PID", () => {
        checkSnapshot("selected:123");
    });

    it("renders correctly with parents", () => {
        const parents = [
            { pid: "foo", title: "Foo" },
            { pid: "bar", title: "Bar" },
        ];
        checkSnapshot("", parents);
    });

    it("allows you to clear a selected PID", async () => {
        render(getPicker("selected:123"));
        await userEvent.setup().click(screen.getByText("Clear"));

        expect(callback).toHaveBeenCalledWith("");
    });

    it("renders correctly without favorite PIDs", () => {
        editorValues.state.favoritePidsCatalog = [];
        checkSnapshot();
    });
});
