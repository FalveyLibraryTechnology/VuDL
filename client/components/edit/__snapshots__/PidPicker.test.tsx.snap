// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PidPicker renders correctly with a selected PID 1`] = `
<Fragment>
  Selected pid: 
  selected:123
  . 
  <button
    onClick={[Function]}
  >
    Clear
  </button>
</Fragment>
`;

exports[`PidPicker renders correctly with favorite PIDs 1`] = `
<Fragment>
  <ForwardRef(Accordion)>
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Enter PID manually
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <label>
        Enter PID:
         
        <input
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <button
        onClick={[Function]}
      >
        Set
      </button>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)>
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Choose PID from Tree
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <Component
        selectCallback={[MockFunction]}
      />
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)>
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Choose PID from Favorites
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <ul>
        <li
          key="favorite_foo:123"
        >
          <button
            onClick={[Function]}
          >
            first test
          </button>
        </li>
        <li
          key="favorite_foo:124"
        >
          <button
            onClick={[Function]}
          >
            second test
          </button>
        </li>
      </ul>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
</Fragment>
`;

exports[`PidPicker renders correctly with parents 1`] = `
<Fragment>
  <ForwardRef(Accordion)>
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Enter PID manually
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <label>
        Enter PID:
         
        <input
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <button
        onClick={[Function]}
      >
        Set
      </button>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)>
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Choose PID from Tree
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <Component
        selectCallback={[MockFunction]}
      />
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)
    key="clone_parent_foo"
  >
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Choose PID from Parent (
      Foo
      )
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <Component
        pid="foo"
        selectCallback={[MockFunction]}
      />
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)
    key="clone_parent_bar"
  >
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Choose PID from Parent (
      Bar
      )
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <Component
        pid="bar"
        selectCallback={[MockFunction]}
      />
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)>
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Choose PID from Favorites
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <ul>
        <li
          key="favorite_foo:123"
        >
          <button
            onClick={[Function]}
          >
            first test
          </button>
        </li>
        <li
          key="favorite_foo:124"
        >
          <button
            onClick={[Function]}
          >
            second test
          </button>
        </li>
      </ul>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
</Fragment>
`;

exports[`PidPicker renders correctly without favorite PIDs 1`] = `
<Fragment>
  <ForwardRef(Accordion)>
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Enter PID manually
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <label>
        Enter PID:
         
        <input
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <button
        onClick={[Function]}
      >
        Set
      </button>
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
  <ForwardRef(Accordion)>
    <ForwardRef(AccordionSummary)
      expandIcon={<Memo />}
    >
      Choose PID from Tree
    </ForwardRef(AccordionSummary)>
    <ForwardRef(AccordionDetails)>
      <Component
        selectCallback={[MockFunction]}
      />
    </ForwardRef(AccordionDetails)>
  </ForwardRef(Accordion)>
</Fragment>
`;
